### ISSUE TITLE
When RStudio is started on an external display, 3/4 of the window is black

### ISSUE THREAD

**Issue by yihui** [rstudio/rstudio#3479]
### System details

```
version  R version 3.5.1 (2018-07-02)
system   x86_64, darwin15.6.0        
ui       RStudio (1.2.985)      
```

### Steps to reproduce the problem

I'm using a Thunderbolt display. If I start RStudio directly on this external display, it ends up in this funny window:

![image](https://user-images.githubusercontent.com/163582/45513622-c4245f80-b768-11e8-94b7-c4f6ec8e4610.png)

I have to drag the window to my laptop's screen, then drag it back to the external display, and it will become normal (occupying the full window). It seems it cannot properly resize itself when starting up on a larger display.

I have noticed this problem for a while. It didn't just appear today.

**Comment by kevinushey** [comment]
I can reproduce this with my in-office thunderbolt monitor as well. I only see it when working on my laptop directly (ie both the laptop retina display and the plain old regular DPI thunderbolt display are being used).

I'll see if I can figure out what's going on.

**Comment by kevinushey** [comment]
Looks like yet another QtWebEngine / Chromium bug. The issue goes away in my case if I switch to software rendering:

![screen shot 2018-09-13 at 4 17 43 pm](https://user-images.githubusercontent.com/1976582/45520946-a1964480-b770-11e8-9bb7-fb2e17e987df.png)

Can you give that a try and let me know if it works?

**Comment by yihui** [comment]
Yes, software rendering works fine.

**Comment by kevinushey** [comment]
This should now 'just work' in the latest daily builds of RStudio (we'll turn on software rendering when we detect this kind of display configuration). Hopefully we'll be able to relax this constraint in a future QtWebEngine update.

I'm going to keep this ticket open since we need to test this with other high DPI (non-Apple, non-Retina) displays.

**Comment by yihui** [comment]
Confirmed in my case. Thanks!

**Comment by kevinushey** [comment]
FWIW I tested this and all looked well on a two high-DPI (but non-retina) display configuration so I'm going to close this unless we hear new reports.

**Comment by kevinushey** [comment]
Re-opening as the original fix we applied was too broad (it also forced software rendering in configurations where it shouldn't have.

Also frustrated to learn that the issue does _not_ reproduce on my new laptop, but did on my old one.

**Comment by jmcphers** [comment]
Anecdotally I've also seen this exact problem (content stuck in 1/4 of the window) on the production version of Chrome (currently 69) in similar situations -- for example, if Chrome is running on my laptop, I plug in a non-Retina display, and Chrome is auto-moved to that display, it has this rendering problem for a second or two until it fixes itself. 

I wonder if rather than dropping into software rendering permanently there's some kind of reload we can perform which will set things straight.

**Comment by kevinushey** [comment]
I can't even reproduce on my old laptop anymore. The only thing I can recall that's changed in the update to Mojave.

@yihui: are you still running High Sierra, or have you updated to Mojave? Do you still see this issue in the latest daily builds?

**Comment by yihui** [comment]
I'm still on High Sierra, and I don't have this issue (currently using v1.2.1045).

**Comment by kevinushey** [comment]
@yihui sorry to trouble you but can you try downloading the latest daily build and testing once more? We had to revert the code we added to originally paper over this issue but I'm surprised that I can no longer reproduce myself.

https://s3.amazonaws.com/rstudio-ide-build/desktop/macos/RStudio-1.2.1064.dmg

**Comment by ateucher** [comment]
I still have this issue with the latest daily build (1.2.1064). I feel like it was gone for a few days but it is back. Using a run of the mill Lenovo display connected to a 2016 15" Macbook Pro with High Sierra (10.13.6).

**Comment by yihui** [comment]
@kevinushey The issue is back in 1.2.1064.

**Comment by jcheng5** [comment]
At @kevinushey's request, I tried reproing with 1.2.1064. It does repro for me on macOS Sierra 10.12.6.

![image](https://user-images.githubusercontent.com/129551/47166747-befb8880-d2b1-11e8-9cbb-434ac04d354d.png)

**Comment by kevinushey** [comment]
I spent a lot of time looking at this today. Here's what I've learned so far:

- It seems only RStudio Desktop is affected by this issue; other Qt example applications are not.
- The issue is fixed in the Qt 5.12.0 beta.
- Trying to 'trick' the window into re-rendering by triggering / firing various window changed / screen changed Qt events doesn't help.
- Toggling the driver blacklist / bug workarounds doesn't make a difference.
- The various GPU / rendering Chromium flags don't seem to help with Qt 5.11.

Unless we want to sink some more time into this I think the fix might ultimately be to take on Qt 5.12 before we ship RStudio v1.2.

**Comment by jmcphers** [comment]
This release cycle we've been on a real treadmill of pulling in Qt upgrades to fix one or two thorny issues that seem totally unsurmountable without an upgrade, only to find (after a week or two of integration work) that there are _another_ one or two thorny issues introduced with the new version.

Qt 5.12 does have the advantage of being an LTS release, although frankly 5.9 was too and we were forced to move on from it pretty quickly since it didn't actually get the bugfixes we needed. What I'd suggest is to wait for the formal release of 5.12 (happening in just a few weeks), then make a 5.12 branch, do a private build on all three desktop platforms, and have QA spend a little dedicated time hammering on it before we consider taking it on master this late in the release.

**Comment by kevinushey** [comment]
On the other hand, we've already committed to pushing the RStudio release back to early next year so we still have a buffer of time to shake out any issues encountered with Qt 5.12. I'm at least somewhat more optimistic that since we've been tracking Qt updates more closely this cycle that moving to Qt 5.12 will be less painful than the less updates have been, but still -- the devil you know, versus the devil you don't...

**Comment by kevinushey** [comment]
Unfortunately shortcuts in text inputs seem to be broken in Qt 5.12 (how fun)

https://bugreports.qt.io/browse/QTBUG-71351

**Comment by kevinushey** [comment]
Qt 5.12 is scheduled for release first week of December; the plan is to produce some candidate builds of RStudio using Qt 5.12 and do some smoke-testing to see if we could safely update to Qt 5.12 (which appears to resolve these issues, at least in some local testing I've done)

**Comment by eipi10** [comment]
I just posted about this issue on [RStudio Community](https://community.rstudio.com/t/rstudio-ide-panes-shrunken-to-only-a-quarter-of-rstudio-application-window-when-opening-any-project/22602) and Jonathan at RStudio pointed me here. I'm experiencing this issue with RStudio Version 1.2.1237 Build 1246 (7a011999) (Preview version) but not with Version 1.1.463 (latest stable release). I'm on a Macbook Pro 15" 2018 running OSX 10.13.6 with two external Dell monitors. 

If I drag the RStudio application window onto one of the other screens, the shrunken screen probolem goes away for the rest of the session (regardless of where I drag the window at any later point), but returns when I close and reopen a project. 

I confirmed that changing to software rendering eliminates the problem.
